text/temp/PALENICA/Larsson_Sadakane_Suffix_Array.tex

```diff
@@ -0,0 +1,105 @@
+\end{enumerate}
+
+Na pierwszy rzut oka można by wyciągnąć wniosek, że algorytm działa w czasie $O(n\ log^2\ n)$. Dokładniejsza analiza pozwala na ograniczenie przez $O(n\ log\ n)$. Na potrzeby analizy zakładamy, że \textit{ternary quicksort} dzieli zawsze w medianie (możliwe do uzyskania przy wydajnym algorytmie znajdowania mediany, nieefektywne w praktyce). Proces sortowania całej tablicy sufiksowej (nie tylko jednego quicksorta) przedstawimy jako ternarne drzewo obliczeń.
+
+\begin{lemma}
+Długość ścieżki od korzenia do liścia jest ograniczona z góry przez $2 log\ n + 3$
+\begin{proof}
+Każdy węzeł `środkowy` na ścieżce odpowiada podwojeniu \textit{h-order} w którym znajduje się rozważana tablica. Takich węzłów może być więc $log\ n + 1$ Z faktu, że używamy mediany po zejściu lewym lub prawym dzieckiem długość rozważanej tablicy zmniejsza się o połowę. Znów takich węzłów będzie maksymalnie $log\ n + 1$
+\end{proof}
+\end{lemma}
+Na danym poziomie drzewa ilość wykonanej pracy będzie $O(n)$ bo to obliczenie odpowiada wykonaniu splita na rozłączne podtablice. Stąd dostajemy złożoność $O(n\ log\ n)$ bo aktualizacja podtablic odpowiedzialnych za grupy również jest wykonywana w czasie liniowym od rozmiaru podtablicy.
+
+W algorytmie autorzy wprowadzają usprawnienia które nie wpływają one negatywnie na czas wykonania. Opiszemy tutaj większość z nich (te które zostały wykorzystane w implementacji).
+
+\textbf{Eliminacja tablicy długości $L$}. Tablica długości dla grup nieposortowanych jest nam zbędna ze względu na to jak numerujemy grupy. Przypomnijmy, numer grupy $I[f \ldots g]$ to $g$, czyli $V[f-1] = g$ a w konsekwencji rozmiar grupy to $V[f-1] - g + 1$. Do spamiętania rozmiarów wspólnych grup posortowanych możemy użyć tablicy $I$. Wpólne grupy posortowane nie pojawią się w wywołaniu quicksorta a tylko tam korzystamy z $I$ w trakcie działania algorytmu. Jedynym problemem pozostaje odzyskanie porządku w $I$ na samym końcu algorytmu, ale tutaj wystarczy, że skorzystamy z tablicy $V$ trzymającej numery grup (indywidualne grupy posortowane mają różne numery), więc $I[V[i]] = i$ dla $0 \leq i < n+1$ odzyska tablicę sufiksową na końcu. Ostatecznie otrzymujemy następującą procedurę odzyskania długości grupy: jeśli $I[i] < 0$ to $I[i \ldots i - I[i] - 1]$ jest wspólną grupą posortowaną. W przeciwnym wypadku $I[i \ldots i + V[I[i]-1]]$ jest nieposortowaną grupą. 
+
+\textbf{Połączenie sortowania i aktualizowania tablic}. Zauważmy, że wykonanie połączenia dwóch następujących wspólnych grup posortowanych możemy wykonywać dopiero przy kroku 4 algorytmu gdy iterujemy się po kolejnych nieposortowanych grupach w $I$ przeskakując nad wspólnymi grupami posortowanymi. Aktualizacja tablic dla grup nieposortowanych w czasie wykonania sortowania musi być wykonana z uwagą na to żeby nie zmienić wartości w $V$ które będą wykorzystane jako klucze sortowania. Fakt wykorzystania quicksorta zapewnia nam, że ustalenie kolejności aktualizacji tablic jest proste:
```

Comment:
"Fakt wykorzystania quicksorta zapewnia nam, że ustalenie kolejności aktualizacji tablic jest proste" - I'm probably having a blackout for not getting this, but could you add a line-two why this works?

And another question: how we were supposed to update $V$ in a consistent way without this improvement? Maybe it's worth mentioning above, just around the basic algorithm outline?

Reply:
Good point. I had issues when reading the article to grasp this point as well. It was explained in the article in similar way and I didn't expand this explanation further but in hindsight - I should have because it's the core of this particular improvement/refinement.

Also, added a few words on possible implementation in basic algorithm version.

Reply:
I'm sorry, but that's still unclear. I know that the original article might be sloppy on that point, but it's the main aim of this project to be comprehensible for others.

Please either expand it even more, provide some toy example or mark with TODO.

Reply:
OK, I guess we finally found more-or-less complete explanation, why the algorithm works. Thank you for the discussion and examples.